src/funcx_common/task_storage/chain.py

```diff
@@ -0,0 +1,55 @@
+import typing as t
+
+from ..tasks import TaskProtocol
+from .base import StorageException, TaskStorage
+
+
+class ChainedTaskStorage(TaskStorage):
+    """
+    Chained storage combines multiple other concrete storage options.
+
+    The storage methods are used on a "first-come, first-served" basis, with the
+    intended usage pattern being
+
+    >>> slow_storage = ...
+    >>> fast_storage = ...
+    >>> task_storage = ChainedTaskStorage(fast_storage, slow_storage)
+
+    In this scenario, the fast storage will be tried first for both the store and get
+    operations.
+
+    If the fast storage rejects a storage operation, the slow storage will be tried. A
+    failure (exception) would be raised immediately, however.
+
+    The assumption here is that the fast storage responds quickly enough that it can be
+    checked first in all cases, even without additional context for the call.
+    """
+
+    storage_id = "ChainedTaskStorage"
+
+    def __init__(self, *storages: TaskStorage) -> None:
+        self._storages = list(storages)
+        self.storage_map = {storage.storage_id: storage for storage in storages}
+
+    def store_result(self, task: TaskProtocol, result: str) -> bool:
+        exception_stack = []
+        for store in self._storages:
+            try:
+                return store.store_result(task, result)
+            except StorageException as e:
+                exception_stack.append(e)
+                pass
+
+        raise StorageException(
+            f"All storage methods failed to store data: {exception_stack}"
+        )
+
+    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
+        # We are special casing here because if the
+        if not task.result and not task.result_reference:
+            return None
+        for store in self._storages:
+            if store.storage_id == task.result_reference["storage_id"]:  # type: ignore
```

Comment:
I think the presence of this type-ignore comment is actually hiding a nefarious bug.

The reason that mypy doesn't type-narrow in this case is that the early exit above, in the if-statement, is using an `and`. That means that it's possible for `task.result_reference` to be `None` when this block is reached.

I think a reasonable resolution would be to make the start of `get_result()` run
```python
if not task.result_reference:
    return None
```

meaning that if the `result_reference` field is not populated, results can never be retrieved via this interface.

But I'm also aware that there's a backwards-compatibility concern we have.

Why not avoid trying to shoehorn this into a generic `ChainedTaskStorage` interface altogether?
Define a new storage, like `LegacyRedisTaskStorage` which takes another storage and checks for `task.result` before even getting into its dispatch logic?

The comment there appears to be partial. I think we're missing information here about the interplay of new, "normalized", task storage and the existing state of tasks which may have `result` populated without `result_reference`.

Reply:
I've restructured this a bit for clarity. I think with the changes we cover all the bases.